design/baremetal-operator/firmware-interface.md

```diff
@@ -0,0 +1,225 @@
+kind: HostFirmwareComponents
+metadata:
+  namespace: host3firmwarecomponents.metal3.io
+spec:
+  updates:
+  - name: bios
+    url: https://myurl.with.firmware.for.bios
+  - name: bmc
+    url: https://myurl.with.firmware.for.bmc
+status:
+  components:
+  - component: bios
+    initialVersion: "v1.0.0"
+    currentVersion: "v1.0.0"
+    lastVersionFlashed: null
+    updatedAt: null
+  - component: bmc
+    initialVersion: "v1.0.5"
+    currentVersion: "v1.0.5"
+    lastVersionFlashed: null
+    updatedAt: null
+  lastUpdated: "2023-10-13T13:34:06Z"
+```
+
+Example of the Resource after applied:
+
+```yaml
+---
+apiVersion: metal3.io/v1alpha1
+kind: HostFirmwareComponents
+metadata:
+  namespace: host3firmwarecomponents.metal3.io
+spec:
+  updates:
+  - name: bios
+    url: https://myurl.with.firmware.for.bios
+  - name: bmc
+    url: https://myurl.with.firmware.for.bmc
+status:
+  components:
+  - component: bios
+    initialVersion: "v1.0.0"
+    currentVersion: "v1.5.0"
+    lastVersionFlashed: "v.1.5.0"
+    updatedAt: "2023-10-13T13:50:06Z"
+  - component: bmc
+    initialVersion: "v1.0.5"
+    currentVersion: "v1.2.0"
+    lastVersionFlashed: "v1.2.0"
+    updatedAt: "2023-10-13T13:50:06Z"
+  updates:
+  - name: bios
+    url: https://myurl.with.firmware.for.bios
+  - name: bmc
+    url: https://myurl.with.firmware.for.bmc
+  lastUpdated: "2023-10-13T13:50:06Z"
+```
+
+#### Fields Description
+
+- `component`: the name of the firmware component
+- `initialVersion`: the initial firmware version of the component, Ironic
+  retrieves this information when creating the BMH and it can't be changed.
+- `currentVersion`: the current firmware version of the component, initially
+  the value will match the one in `initialVersion`, unless there was a
+  firmware update for the BMH.
+- `lastVersionFlashed`: the last firmware version of the component that was
+  flashed in the BMH, this field will only have a value when a firmware update
+  is executed.
+- `updatedAt`: when the firmware component information was updated by Ironic.
+
+### Risks and Mitigations
+
+- In case of failure when executing the firmware update, BMO will put the BMH
+  in a failed state.
+- New firmware may have fixed a few bugs, but it can also introduce new ones.
+  We will try to mitigate this by providing some versions of tested firmware
+  when possible.
+
+### Work Items
+
+BMO
+
+- Add new CRD for `HostFirmwareComponents`
+- Get the firmware information from Ironic API at the end of the Registration
+  state and store all the information in the `components`
+- Check for changes to `updates` and when detected, call the Ironic
+  clean-steps API
+
+### Dependencies
+
+- Ironic (support already exists)
+- Gophercloud (support already exists)
+
+### Test Plan
+
+- Test on running cluster
+- Verify that updates can be executed on two type of hardware that supports
+  Redfish.
+- Verify that the information about the desired version is available after
+  cleaning.
```

Comment:
BMO has new e2e tests (check the commit history). That allows us to test firmware upgrades against sushy-tools. WDYT?

Reply:
Hummm that could be a thing, but sushy-tools currently doesn't have the capability to execute firmware updates.